### ISSUE TITLE
Group GE windows in the taskbar

### ISSUE THREAD

**Issue by mstv** [gitextensions/gitextensions#6114]
## Current behaviour
Windows of one GE instance appear as if they belong to different applications.

## Expected behaviour
The GE windows' taskbar icons are grouped together.
![grafik](https://user-images.githubusercontent.com/36601201/51207106-2c976f00-190a-11e9-83a2-20e0c8ccab02.png)

## Steps to reproduce
- open GE
- drag the taskbar icon in front of some other application
- open the History window

## Screenshots
![grafik](https://user-images.githubusercontent.com/36601201/51206940-b135bd80-1909-11e9-8e39-e8f1b0462e4d.png)

## Did this work in previous version of GitExtensions
Yes, in 2.51.05

## Environment
- Git Extensions 3.00.00.4433 and 3.01.00.0
- Build 5a313988bf4ab8e0c37bdf1f2c658c9469e1e257
- Git 2.20.1.windows.1
- Microsoft Windows NT 6.1.7601 Service Pack 1
- .NET Framework 4.7.2117.0
- DPI 96dpi (no scaling)

**Comment by gerhardol** [comment]
If I group apps in the taskbar, both 2.x and 3.0 is grouped, they always appear together if the same binary for me.

Otherwise, has this to do with the updated taskbar functionality with notification dot and quick actions?
(Probably #5087 related.)
![image](https://user-images.githubusercontent.com/6248932/51208336-aaa94500-190d-11e9-9f86-dc1cb645bcdc.png)

**Comment by mstv** [comment]
I use the setting "never group". The screenshot for the expected behavior was taken of 2.51.05.

**Comment by gerhardol** [comment]
I normally use "Group when full", but tested Never too.
Same for me in 2.51 and build from latest master.
Any difference in how you start GE? pinning?

**Comment by mstv** [comment]
It depends on the setting "Open last working directory on startup"! I have this on always because I do not need and use the dashboard.
![grafik](https://user-images.githubusercontent.com/36601201/51265256-aa19b880-19b8-11e9-9a9b-260fdc5ce5cf.png)

**Comment by gerhardol** [comment]
> It depends on the setting "Open last working directory on startup"! 

Hm, same setting for me

**Comment by mstv** [comment]
Today, I re-pinned the release build to the taskbar and it works.
The unpinned debug build's History window gets grouped with the pinned _release_ build.

On another Win7 machine where I have never pinned GE to the taskbar, several builds behave differently in dependency on [not] using the dashboard.

In a Win10 VM, the history appears separately in the taskbar.
There the setting "Open last working directory on startup" has no effect: no dashboard on startup.

**Comment by vbjay** [comment]
https://docs.microsoft.com/en-us/windows/desktop/shell/taskbar-extensions

**Comment by gerhardol** [comment]
> The unpinned debug build's History window gets grouped with the pinned release build.

The grouping is always per binary for me, if I hover as I show above, each release is together.
I add shortcuts to the toolbar, similar to pinning for releases. Up until last weeks build, they appeared over the icon for the release build. Now they are separate, not sure what it is based on. 

Is there anything to improve from this issue?

**Comment by vbjay** [comment]
We could see if we can alter the ID by repo and a special ID for dashboard. Maybe even debug-repo and release-repo.

**Comment by vbjay** [comment]
https://docs.microsoft.com/en-us/windows/desktop/shell/appids

**Comment by vbjay** [comment]
Looks like the api code pack allows us to manage this.

**Comment by gerhardol** [comment]
Should we group Browse and File Viewer windows?

**Comment by vbjay** [comment]
I think we should group process wide.  Lets say we open debug and release of the GE repo.  We would have two GE buttons in the taskbar.  All windows of the debug process would be in one button.  

Let's say you open the debug twice of the GE repo.  Still two buttons.  One of  the debug and one of the release.  You would still have two browse windows in the debug group.

**Comment by vbjay** [comment]
https://github.com/aybe/Windows-API-Code-Pack-1.1/blob/ae73c1294fe9d47c5052d090b945f69a6364e3a8/source/WindowsAPICodePack/Shell/Taskbar/TaskbarManager.cs#L205

**Comment by vbjay** [comment]
Change repo, change id.

**Comment by vbjay** [comment]
If possible, we should also have command buttons per repo.  
![image](https://user-images.githubusercontent.com/3629489/51286234-16e48100-19c0-11e9-8132-6aef4a8c8262.png)

**Comment by mstv** [comment]
> Change repo, change id.

I'm afraid, then pinning GE to the taskbar would not work well anymore.
I want to pin the release version as the fourth taskbar icon. Then I can quickly switch to current repo by hitting `Win+4`.

**Comment by mstv** [comment]
I understand the use case with app id per repo. Though I would not use it. I guess / I'm afraid I would have to switch it off because a pinned release version is important to me, i.e. grouping by executable.
When I test something new I run the Debug build (which is not pinned).
I switch (sub-)repos using the toolbar: ![grafik](https://user-images.githubusercontent.com/36601201/51350346-3a82f680-1aa8-11e9-88f5-6080b506fbe2.png).

**Comment by vbjay** [comment]
Well I've seen debug meld into my pinned one.   So I think us making sure
debug/release was part of the appid, would help. I think an option to group
by repo would be useful.

**Comment by mstv** [comment]
Yes, I did. It will be really good to use the superproject's id for the submodule repos.
(I switch to submodules using the submarine toolbar button.)

**Comment by vbjay** [comment]
I do too but ur us not the only way. It spins up a new browse window.

**Comment by gerhardol** [comment]
My summary (just expanding on vbjay):
As a developer, I have many versions pinned/added to the toolbar and would prefer to switch on binary first, not just Debug/Release. But that is that first level only.

For second level switching on superproject would be a good functionality when thinking of it. I often open a group of Browse for a "project".

For the third level, all Windows related to a project (Browse, Browse open from Browse, File History etc) should be grouped with the "project".

This kindof assumes that application id can be changed dynamically, if opening a repo not in the current "project", the id should change and grouping change.

**Comment by vbjay** [comment]
From my research it looks like we can. I just don't know if jumplists would
be affected.   API code pack should it us do this.
